contracts/utils/packing.cairo

```diff
@@ -1,157 +1,143 @@
-    let new_row = cell_binary + row_to_store
-
-    return (new_row)
 end
 
-
-# Pre-sim. Walk columns for a given row and saves state to an array.
-func append_cols{
+func pack_game{
         syscall_ptr : felt*,
         bitwise_ptr : BitwiseBuiltin*,
         pedersen_ptr : HashBuiltin*,
         range_check_ptr
     }(
-        cell_states : felt*,
-        row : felt,
-        col : felt,
-        stored_row : felt*
+        cells_len : felt,
+        cells : felt*
+    ) -> (
+        packed_game
     ):
     alloc_locals
-    if col == 0:
-        return ()
-    end
 
-    append_cols(cell_states=cell_states,
-        row=row, col=col-1, stored_row=stored_row)
-    # (Note, on first entry, col=1 so col-1 gets the index)
-    #local pedersen_ptr : HashBuiltin* = pedersen_ptr
-    #local cell_states : felt* = cell_states
-    #local bitwise_ptr : BitwiseBuiltin* = bitwise_ptr
-    # state = 2**column_index AND row_binary
-    # Column zero is the MSB, so (DIM_index - col_index) accesses the bit.
-    let binary_position = (DIM - 1) - (col - 1)
-    # let (mask) = pow(2, binary_position)
-    # Avoid bitwise builtin use ecause it has a global cap ~1024.
-    # let (state) = bitwise_and(stored_row, mask)
-    let state = stored_row[binary_position]
-    # E.g., if in col_index 2, for an alive 'state=4 (0b100)',
-    # convert to 1
-    local alive_or_dead
-    if state == 0:
-        assert alive_or_dead = 0
-    else:
-        assert alive_or_dead = 1
-    end
-    let index = row * DIM + (col - 1)
-    assert cell_states[index] = alive_or_dead
-
-
-    return ()
+    let (low) = pack_cells(
+        cells_len=LOW_ARRAY_LEN,
+        cells=cells,
+        index=0,
+        power=1,
+        packed_cells=0
+    )
+    let (high) = pack_cells(
+        cells_len=HIGH_ARRAY_LEN,
+        cells=cells + LOW_ARRAY_LEN,
+        index=0,
+        power=1,
+        packed_cells=0
+    )
+    let packed_game = high * SHIFT + low
+
+    return(packed_game)
 end
 
-# Performs bitwise_and functionality without using bitwise builtin.
-func custom_bitwise_and{
+func pack_single_cell{
         syscall_ptr : felt*,
         bitwise_ptr : BitwiseBuiltin*,
         pedersen_ptr : HashBuiltin*,
         range_check_ptr
     }(
-        packed_value : felt,
-        binary_position : felt
+        cell_index : felt,
+        current_state : felt
     ) -> (
-        res : felt
+        packed_game : felt
     ):
     alloc_locals
-    let (local result_array : felt*) = alloc()
-    # Only read up to 32 values.
-    split_int(value=packed_value, n=32, base=2, bound=2, output=result_array)
-    let res = result_array[binary_position]
-    return (res)
+
+    let (high, low) = split_felt(current_state)
+    # determines on which array ("part") the index is placed
+    let (part) = is_le_felt(cell_index, LOW_ARRAY_LEN - 1)
+
+    # Cell lies in the upper part of the board, which is encoded
+    # in "low" part of the felt
```

Comment:
Are you sure about the low and high in those comments? Last cell is last bit of the lower. The logic itself is ok.

Reply:
It's actually the other way around - last cell is the last bit of the higher. So for example, if the board has only one live cell at 0 index, the whole game is encoded as 1 (and not as 2**128).